PortalSwift/Sources/PortalSwift/Storage/ICloudStorage.swift

```diff
@@ -0,0 +1,116 @@
+//
+//  ICloudStorage.swift
+//
+//  Created by Portal Labs, Inc.
+//  Copyright © 2022 Portal Labs, Inc. All rights reserved.
+//
+
+import CommonCrypto
+import Foundation
+
+/// A storage class that uses iCloud's key-value store to store/retrieve private keys.
+public class ICloudStorage: Storage, PortalStorage {
+  public enum ICloudStorageError: Error {
+    case noAPIKeyProvided(String)
+    case noAccessToICloud(String)
+    case notSignedIntoICloud(String)
+    case unableToDeriveKey(String)
+    case unableToRetrieveClient(String)
+    case failedValidateOperations(String)
+    case unknownError
+  }
+
+  public enum ICloudStatus: String {
+    case available
+    case notSignedIn
+    case noAccess
+  }
+
+  /// The Portal API instance to retrieve the client's and custodian's IDs.
+  public var api: PortalApi?
+  /// The key used to store the private key in iCloud.
+  public var key: String = ""
+
+  public let encryption: PortalEncryption
+
+  private let isSimulator = TARGET_OS_SIMULATOR != 0
+  private let storage: PortalKeyValueStore
+
+  /// Initializes a new ICloudStorage instance.
+  public init(
+    encryption: PortalEncryption? = nil,
+    keyValueStore: PortalKeyValueStore? = nil
+  ) {
+    self.encryption = encryption ?? PortalEncryption()
+    self.storage = keyValueStore ?? PortalKeyValueStore()
+  }
+
+  /*******************************************
+   * Public functions
+   *******************************************/
+
+  public func delete() async throws -> Bool {
+    let key = try await getKey()
+    return self.storage.delete(key)
+  }
+
+  public func read() async throws -> String {
+    let key = try await getKey()
+    return self.storage.read(key)
+  }
+
+  public func write(_ value: String) async throws -> Bool {
+    let key = try await getKey()
+    return self.storage.write(key, value: value)
+  }
+
+  public func validateOperations() async throws -> Bool {
```

Comment:
I see this function used to use helper functions that the above functions where not using like `rawWrite` etc. So now all of it is consistent and using the same methods in PortalKeyValueStore

Reply:
Yep! So it actually validates the behavior of what we'll be doing if it succeeds 👌